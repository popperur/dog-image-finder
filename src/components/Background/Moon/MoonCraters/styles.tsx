import styled, { keyframes } from "styled-components";

export const MoonCraterContainer = styled.div``;

export const MoonCraterHolder = styled.div<{ $styles?: string[] }>`
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  ${props => props.$styles && props.$styles.join(";")};
`;

export const MoonSignalPlayerContainer = styled.div`
  position: relative;
  top: 54px;
  left: 17px;
  color: #888;
`;

const moonSignalPlayerPulseFrames = keyframes`
  0% {
    opacity: 1;
  }
  50% {
    opacity: 0.5;
  }
  100% {
    opacity: 1;
  }
`;

export const MoonSignalPlayerButton = styled.div.attrs({
  role: "button",
  tabIndex: 0,
})`
  cursor: pointer;
  width: 20px;
  animation: ${moonSignalPlayerPulseFrames} 2s infinite;
`;
