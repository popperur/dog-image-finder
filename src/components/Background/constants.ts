export const TOOLTIP_PAUSED =
  "Radio signal from the Moon! This incredible discovery could greatly enhance your experience in browsing for dogs!";
export const TOOLTIP_PLAYING = "Auuuuuu!";

// top, left, width, height, animationDelay
export const STARS = [
  [45, 66, 6, 6, 2],
  [70, 48, 5, 5, 4],
  [4, 44, 3, 3, 1],
  [23, 53, 1, 1, 2],
  [26, 93, 3, 3, 1.3],
  [46, 13, 4, 4, 2.3],
  [66, 34, 2, 2, 1.3],
  [36, 3, 3, 3, 3.3],
  [76, 54, 1, 1, 2.3],
  [56, 24, 3, 3, 1.3],
  [76, 25, 3, 3, 0.3],
  [56, 55, 1, 1, 2.31],
  [36, 35, 3, 3, 1.31],
  [86, 45, 2, 2, 2.31],
  [82, 85, 3, 3, 0.31],
  [81, 15, 1, 1, 1.31],
  [83, 35, 3, 3, 2.31],
  [23, 55, 2, 2, 4],
  [53, 50, 1, 1, 1.31],
  [83, 20, 1, 1, 1.51],
];

// top, left, width, height, opacity
export const MOON_CRATERS = [
  [23, 33, 17, 17, 0.5],
  [48, 10, 30, 30, 0.7],
  [90, 14, 20, 20, 0.4],
  [18, 63, 25, 25, 0.6],
  [33, 105, 10, 10, 0.8],
  [82, 55, 14, 14, 0.9],
  [122, 34, 20, 20, 0.6],
  [102, 84, 16, 16, 0.8],
  [63, 113, 20, 20, 0.9],
  [116, 119, 12, 12, 0.6],
];

// top, left, width, height, animationDuration, animationDelay
export const MOON_GLOWS = [
  [20, 30, 260, 260, 3, 0],
  [-20, -10, 340, 340, 3, 0.5],
];

// bottom, left, width, height
export const HILLS = [
  [183, 120, 20, 239],
  [259, 229, 28, 255],
  [211, 334, 28, 246],
];
