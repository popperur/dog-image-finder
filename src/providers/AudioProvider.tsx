import React, { useReducer, useRef } from "react";
import { audioReducer, AudioContext } from "context/AudioContext.ts";

interface AudioProviderProps {
  children: React.ReactNode;
}

function AudioProvider({ children }: AudioProviderProps) {
  const [state, dispatch] = useReducer(audioReducer, { isPlaying: false });
  const audioRef = useRef<HTMLAudioElement | null>(null);

  const play = () => {
    audioRef.current?.play();
    dispatch({ type: "PLAY" });
  };

  const pause = () => {
    audioRef.current?.pause();
    dispatch({ type: "PAUSE" });
  };

  return (
    <AudioContext.Provider value={{ state, play, pause }}>
      {children}
      <audio ref={audioRef} role="audio">
        <source src="/audio/moon_signal.mp3" type="audio/mpeg" />
      </audio>
    </AudioContext.Provider>
  );
}

export default AudioProvider;
