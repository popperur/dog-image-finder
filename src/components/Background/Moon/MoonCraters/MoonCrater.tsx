import { MoonCraterHolder } from "components/Background/Moon/MoonCraters/styles.tsx";

interface MoonCraterProps {
  top: number;
  left: number;
  width: number;
  height: number;
  opacity: number;
}

function MoonCrater({ top, left, width, height, opacity }: MoonCraterProps) {
  const stylesArray = [
    `top: ${top}px`,
    `left: ${left}px`,
    `width: ${width}px`,
    `height: ${height}px`,
    `opacity: ${opacity}`,
  ];

  return <MoonCraterHolder $styles={stylesArray} />;
}

export default MoonCrater;
