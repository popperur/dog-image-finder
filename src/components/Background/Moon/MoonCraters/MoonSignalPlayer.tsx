import { useState } from "react";
import { SoundFilled, SoundOutlined } from "@ant-design/icons";
import { Tooltip } from "antd";
import { MoonSignalPlayerButton } from "components/Background/Moon/MoonCraters/styles.tsx";
import {
  TOOLTIP_PAUSED,
  TOOLTIP_PLAYING,
} from "components/Background/constants.ts";
import useAudio from "hooks/useAudio.ts";

function MoonSignalPlayer() {
  const { state, play, pause } = useAudio();
  const [tooltipTitle, setTooltipTitle] = useState(
    state.isPlaying ? TOOLTIP_PLAYING : TOOLTIP_PAUSED,
  );

  const toggleMusic = () => {
    setTooltipTitle(state.isPlaying ? TOOLTIP_PAUSED : TOOLTIP_PLAYING);
    state.isPlaying ? pause() : play();
  };

  return (
    <>
      <Tooltip placement="left" title={tooltipTitle} color="#0f1721">
        <MoonSignalPlayerButton onClick={toggleMusic} aria-label="toggle music">
          {state.isPlaying ? (
            <SoundFilled aria-label="SoundFilledIcon" />
          ) : (
            <SoundOutlined aria-label="SoundOutlinedIcon" />
          )}
        </MoonSignalPlayerButton>
      </Tooltip>
    </>
  );
}

export default MoonSignalPlayer;
