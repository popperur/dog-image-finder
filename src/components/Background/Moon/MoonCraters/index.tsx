import {
  MoonCraterContainer,
  MoonSignalPlayerContainer,
} from "components/Background/Moon/MoonCraters/styles.tsx";
import MoonSignalPlayer from "components/Background/Moon/MoonCraters/MoonSignalPlayer.tsx";
import { MOON_CRATERS } from "components/Background/constants.ts";
import MoonCrater from "components/Background/Moon/MoonCraters/MoonCrater.tsx";

function MoonCraters() {
  const moonCratersProps = MOON_CRATERS.map(
    ([top, left, width, height, opacity]) => ({
      top,
      left,
      width,
      height,
      opacity,
    }),
  );

  return (
    <MoonCraterContainer>
      {moonCratersProps.map((moonCraterProps, index) => (
        <MoonCrater key={index} {...moonCraterProps} />
      ))}
      <MoonSignalPlayerContainer>
        <MoonSignalPlayer />
      </MoonSignalPlayerContainer>
    </MoonCraterContainer>
  );
}

export default MoonCraters;
