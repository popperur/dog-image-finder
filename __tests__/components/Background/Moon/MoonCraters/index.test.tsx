import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import MoonCraters from "components/Background/Moon/MoonCraters";
import { MOON_CRATERS } from "components/Background/constants.ts";
import AudioProvider from "providers/AudioProvider.tsx";

describe("MoonCraters component", () => {
  it("matches the snapshot", () => {
    const { asFragment } = render(
      <AudioProvider>
        <MoonCraters />
      </AudioProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows all craters and the moon signal player", () => {
    const { container } = render(
      <AudioProvider>
        <MoonCraters />
      </AudioProvider>,
    );

    expect(container.firstChild?.childNodes).toHaveLength(
      MOON_CRATERS.length + 1 /* moon signal player */,
    );
  });
});
