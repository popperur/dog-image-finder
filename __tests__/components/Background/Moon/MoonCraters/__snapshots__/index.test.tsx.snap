// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MoonCraters component > matches the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  top: 23px;
  left: 33px;
  width: 17px;
  height: 17px;
  opacity: 0.5;
}

.c1 {
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  top: 48px;
  left: 10px;
  width: 30px;
  height: 30px;
  opacity: 0.7;
}

.c2 {
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  top: 90px;
  left: 14px;
  width: 20px;
  height: 20px;
  opacity: 0.4;
}

.c3 {
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  top: 18px;
  left: 63px;
  width: 25px;
  height: 25px;
  opacity: 0.6;
}

.c4 {
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  top: 33px;
  left: 105px;
  width: 10px;
  height: 10px;
  opacity: 0.8;
}

.c5 {
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  top: 82px;
  left: 55px;
  width: 14px;
  height: 14px;
  opacity: 0.9;
}

.c6 {
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  top: 122px;
  left: 34px;
  width: 20px;
  height: 20px;
  opacity: 0.6;
}

.c7 {
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  top: 102px;
  left: 84px;
  width: 16px;
  height: 16px;
  opacity: 0.8;
}

.c8 {
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  top: 63px;
  left: 113px;
  width: 20px;
  height: 20px;
  opacity: 0.9;
}

.c9 {
  position: absolute;
  background-color: #cadad6;
  border-radius: 50%;
  top: 116px;
  left: 119px;
  width: 12px;
  height: 12px;
  opacity: 0.6;
}

.c10 {
  position: relative;
  top: 54px;
  left: 17px;
  color: #888;
}

.c11 {
  cursor: pointer;
  width: 20px;
  animation: jDLGYP 2s infinite;
}

<div
    class=""
  >
    <div
      class="c0"
    />
    <div
      class="c1"
    />
    <div
      class="c2"
    />
    <div
      class="c3"
    />
    <div
      class="c4"
    />
    <div
      class="c5"
    />
    <div
      class="c6"
    />
    <div
      class="c7"
    />
    <div
      class="c8"
    />
    <div
      class="c9"
    />
    <div
      class="c10"
    >
      <div
        aria-label="toggle music"
        class="c11"
        role="button"
        tabindex="0"
      >
        <span
          aria-label="SoundOutlinedIcon"
          class="anticon anticon-sound"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="sound"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M625.9 115c-5.9 0-11.9 1.6-17.4 5.3L254 352H90c-8.8 0-16 7.2-16 16v288c0 8.8 7.2 16 16 16h164l354.5 231.7c5.5 3.6 11.6 5.3 17.4 5.3 16.7 0 32.1-13.3 32.1-32.1V147.1c0-18.8-15.4-32.1-32.1-32.1zM586 803L293.4 611.7l-18-11.7H146V424h129.4l17.9-11.7L586 221v582zm348-327H806c-8.8 0-16 7.2-16 16v40c0 8.8 7.2 16 16 16h128c8.8 0 16-7.2 16-16v-40c0-8.8-7.2-16-16-16zm-41.9 261.8l-110.3-63.7a15.9 15.9 0 00-21.7 5.9l-19.9 34.5c-4.4 7.6-1.8 17.4 5.8 21.8L856.3 800a15.9 15.9 0 0021.7-5.9l19.9-34.5c4.4-7.6 1.7-17.4-5.8-21.8zM760 344a15.9 15.9 0 0021.7 5.9L892 286.2c7.6-4.4 10.2-14.2 5.8-21.8L878 230a15.9 15.9 0 00-21.7-5.9L746 287.8a15.99 15.99 0 00-5.8 21.8L760 344z"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
  <audio
    role="audio"
  >
    <source
      src="/audio/moon_signal.mp3"
      type="audio/mpeg"
    />
  </audio>
</DocumentFragment>
`;
